### PR TITLE
Proxy restart fixes

### LINKED ISSUE #6945: SIGHUP of proxy results in TLS errors until original process exits.
### Description

**What happened**: In order to avoid interrupting sessions traversing the proxy we `systemctl reload teleport` regularly and have observed client connectivity/TLS issues until the process in graceful shutdown is terminated.
```
[PROXY:SER] ERRO "proxy2021/05/17 19:05:24 http: TLS handshake error from 192.168.200.1:54962: cache is closed\n" utils/cli.go:304
```

**What you expected to happen**: Clients should only hit the newly forked teleport process and not the process in graceful shutdown which is waiting for client sessions to terminate.

### Reproduction Steps
As minimally and precisely as possible, describe step-by-step how to reproduce the problem.
1. `systemctl reload teleport` on proxy
2. `curl https://proxy:3080/webapi/ping`  observe internal tls error messages or see `tsh login` output in client section below, and `cache is closed` errors in proxy logs.
3. Terminate "old" teleport process and intermittent TLS errors cease.

### Server Details
- Teleport version (run `teleport version`): Tested on 6.1.2 and 6.1.5
- Server OS (e.g. from `/etc/os-release`): 
```
NAME="Red Hat Enterprise Linux Server"
VERSION="7.7 (Maipo)"
ID="rhel"
ID_LIKE="fedora"
VARIANT="Server"
VARIANT_ID="server"
VERSION_ID="7.7"
PRETTY_NAME="Red Hat Enterprise Linux Server 7.7 (Maipo)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:7.7:GA:server"
HOME_URL="https://www.redhat.com/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 7"
REDHAT_BUGZILLA_PRODUCT_VERSION=7.7
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="7.7"
```
- Where are you running Teleport? (e.g. AWS, GCP, Dedicated Hardware): Dedicated hardware

### Client Details
- Tsh version (`tsh version`): 6.1.2 and 6.1.5
- Computer OS (e.g. Linux, macOS, Windows): Linux
- Browser version (for UI-related issues): All browser versions result in SSL protocol errors.
- Installed via (e.g. apt, yum, brew, website download): yum
- Additional details:
```
$ tsh login --proxy=proxy.example.com --auth=azure_ad
ERROR: Get "https://proxy.example.com:3080/webapi/ping/azure_ad": remote error: tls: internal error
```

### Debug Logs
Unfortunately I don't have debug logs from the recent occurrences but will reproduce this today/tomorrow and update with some logs if necessary.

### PR DESCRIPTION
**Pull request by espadolini**
This PR fixes #6945 by ensuring that we close the proxy listeners at the very beginning of the `"proxy.shutdown"` callback; this is necessary because a half-shutdown proxy server cannot function properly, and during rotations or restarts two loops would be accepting incoming TCP connections, even though the loop serviced by the old proxy is likely not able to service requests correctly.

In addition, this fixes a potential race condition that would result in a stuck healthz response: a db service with no statically configured databases will emit a `TeleportOKEvent` at startup, and with a single 50ms wait I was able to make that happen before the `readyz.monitor` function started listening for events, resulting in no more OK events being parsed.

Fixes #6945.

### CODE REVIEW

**Comment by espadolini** [comment]
friendly ping @r0mant @Tener

**Comment by Tener** [comment]
I have skimmed through the changes and they look reasonable given the description, yet I find it hard to build a certainty about the outlined goals as the restart scenarios are involved, so the changes are not "obviously correct".

One thing that stands out is that no changes are made to the tests. How hard would it be to add some coverage here? In an ideal world, this would give us a set of simple tests which fail prior to these changes and passes afterward.

**Comment by espadolini** [comment]
@Tener I've added test coverage for both the immediate bugs fixed in this PR; it's not exhaustive, but more comprehensive tests will be added in the future as I'm currently working on detailing what the current and desired behavior is with regards to restarts (including rotations) and shutdowns vs. connectivity. PTAL?

File: api/utils/sshutils/chconn.go [hunk 1]
```diff
@@ -125,11 +125,11 @@ func (c *ChConn) RemoteAddr() net.Addr {
 // Read reads from the channel.
 func (c *ChConn) Read(data []byte) (int, error) {
 	n, err := c.reader.Read(data)
-	// A lot of code relies on "use of closed network connection" error to
+	// A lot of code relies on net.ErrClosed or "use of closed network connection" error to
 	// gracefully handle terminated connections so convert the closed pipe
 	// error to it.
 	if err != nil && err == io.ErrClosedPipe {
-		return n, trace.ConnectionProblem(err, constants.UseOfClosedNetworkConnection)
+		return n, trace.Wrap(net.ErrClosed, constants.UseOfClosedNetworkConnection)
```
Comment (Tener) [inline api/utils/sshutils/chconn.go:132]: Are we abandoning the use of `trace.ConnectionProblem`? If so, we may want to drop it from the `trace` library or deprecate it. Notably, `trace.ConvertSystemError` may still produce it.

File: lib/utils/errors.go [hunk 2]
```diff
@@ -18,21 +18,34 @@ package utils
...
 	"github.com/gravitational/teleport/api/constants"
 	"github.com/gravitational/trace"
 )
 
-// IsUseOfClosedNetworkError returns true if the specified error
-// indicates the use of closed network connection
-// TODO(dmitri): replace in go1.16 with `errors.Is(err, net.ErrClosed)`
+// IsUseOfClosedNetworkError returns true if the specified error indicates the
+// use of a closed network connection. If the error is a trace.Aggregate, all
+// the errors must be a closed network error.
```
Comment (Tener) [inline lib/utils/errors.go:31]: I think this particular behaviour against `trace.Aggregate` will be surprising. `error.Is` is closer to `Any()` than `All()`.

**Comment by Tener** [comment]
> @Tener I've added test coverage for both the immediate bugs fixed in this PR; it's not exhaustive, but more comprehensive tests will be added in the future as I'm currently working on detailing what the current and desired behavior is with regards to restarts (including rotations) and shutdowns vs. connectivity. PTAL?

Are these planned tests going to be submitted under this PR or a new one?

**Comment by espadolini** [comment]
> Are these planned tests going to be submitted under this PR or a new one?

A new PR, together with whatever changes we implement as we move closer towards the behavior we want out of restarts.

File: lib/utils/errors.go (hunk 2 above)
Comment (espadolini) [reply lib/utils/errors.go:31]: Of the other two "aggregate error" types that I could find, one will do the `Any` behavior while the other will do the check on whatever the first error is; the behavior as I coded this was guided by the fact that in a lot of places we use `IsUseOfClosedNetworkError` as a check for "should we log this or can we ignore it?", so requiring that all the errors are ignorable felt more appropriate. Should we move the aggregate check to `IsOKNetworkError` instead?

File: api/utils/sshutils/chconn.go (hunk 1 above)
Comment (espadolini) [reply api/utils/sshutils/chconn.go:132]: I'm going to restore some of those `trace.ConnectionProblem`s but making sure that they contain a `net.ErrClosed` (or something that contains it), as the check for `errors.Is(err, net.ErrClosed)` would succeed anyway.

I'm not sure of the semantics of `ConnectionProblem`, however: is an error on `(net.Listener).Accept()` on a closed listener really a *connection* problem?

File: lib/utils/errors.go (hunk 2 above)
Comment (espadolini) [reply lib/utils/errors.go]: I moved the check for `trace.Aggregate` to `IsOKNetworkError`, as that one is only used to decide whether or not to log errors, AFAICT.

File: api/utils/sshutils/chconn.go (hunk 1 above)
Comment (espadolini) [reply api/utils/sshutils/chconn.go]: Turned some of these into `ConnectionProblem(net.ErrClosed, "message")` and left alone `chconn.go` - it's handled by the check on the error message, anyway.

**Review by Tener — APPROVED** [review]

**Comment by espadolini** [comment]
friendly ping @r0mant?

File: lib/cache/cache.go [hunk 3]
```diff
@@ -350,10 +350,6 @@ type Cache struct {
 	// fails.
 	initErr error
 
-	// wrapper is a wrapper around cache backend that
-	// allows to set backend into failure mode,
-	// intercepting all calls and returning errors instead
-	wrapper *backend.Wrapper
```
Comment (r0mant) [inline lib/cache/cache.go:356]: Looks like this wrapper is only being used in cache_test.go now. Is it still needed or can it be removed completely?

File: lib/multiplexer/multiplexer.go [hunk 4]
```diff
@@ -191,14 +178,15 @@ func (m *Mux) Serve() error {
...
+		if utils.IsUseOfClosedNetworkError(err) {
+			<-m.context.Done()
 			return nil
 		}
 		select {
-		case <-backoffTimer.C:
-			m.Debugf("backoff on accept error: %v", trace.DebugReport(err))
 		case <-m.context.Done():
 			return nil
+		case <-time.After(5 * time.Second):
```
Comment (r0mant) [inline lib/multiplexer/multiplexer.go:188]: What was the problem with using a ticker? Doesn't calling `time.After` in a loop repeatedly cause some memory leaks? I recall something like this from before.

File: lib/multiplexer/multiplexer.go [hunk 5]
```diff
@@ -191,14 +178,15 @@ func (m *Mux) Serve() error {
 			go m.detectAndForward(conn)
 			continue
 		}
-		if m.isClosed() {
+		if utils.IsUseOfClosedNetworkError(err) {
+			<-m.context.Done()
```
Comment (r0mant) [inline lib/multiplexer/multiplexer.go:182]: Can this deadlock? I'm always a little wary when there's a receive from channel without select. Would it make sense to add some timeout here?

File: lib/multiplexer/multiplexer.go [hunk 6]
```diff
@@ -191,14 +178,15 @@ func (m *Mux) Serve() error {
...
 			return nil
 		}
 		select {
-		case <-backoffTimer.C:
-			m.Debugf("backoff on accept error: %v", trace.DebugReport(err))
 		case <-m.context.Done():
 			return nil
+		case <-time.After(5 * time.Second):
+			m.Debugf("backoff on accept error: %v", trace.DebugReport(err))
```
Comment (r0mant) [inline lib/multiplexer/multiplexer.go:189]: Nit:
```suggestion
			m.WithError(err).Debugf("Backoff on accept error.")
```

File: lib/multiplexer/multiplexer.go [hunk 7]
```diff
@@ -191,14 +178,15 @@ func (m *Mux) Serve() error {
 			go m.detectAndForward(conn)
 			continue
 		}
-		if m.isClosed() {
+		if utils.IsUseOfClosedNetworkError(err) {
```
Comment (r0mant) [inline lib/multiplexer/multiplexer.go:181]: FYI there's also `utils.IsOKNetworkError` which handles a few more scenarios of when a connection is closed somewhat expectedly.

File: lib/multiplexer/tls.go [hunk 8]
```diff
@@ -121,13 +118,14 @@ func (l *TLSListener) Serve() error {
...
+		if utils.IsUseOfClosedNetworkError(err) {
+			<-l.context.Done()
+			return trace.Wrap(err, "listener is closed")
 		}
 		select {
-		case <-backoffTimer.C:
 		case <-l.context.Done():
-			return trace.ConnectionProblem(nil, "listener is closed")
+			return trace.Wrap(net.ErrClosed, "listener is closed")
+		case <-time.After(5 * time.Second):
```
Comment (r0mant) [inline lib/multiplexer/tls.go:128]: Same note as above, I'd keep the ticker to be on a safe side.

File: lib/multiplexer/web.go [hunk 9]
```diff
@@ -96,19 +95,18 @@ func (l *WebListener) DB() net.Listener {
...
+				<-l.context.Done()
+				return trace.Wrap(err, "listener is closed")
 			}
 			select {
-			case <-backoffTimer.C:
-				l.log.WithError(err).Warn("Backoff on network error.")
 			case <-l.context.Done():
-				return trace.ConnectionProblem(nil, "listener is closed")
+				return trace.Wrap(net.ErrClosed, "listener is closed")
+			case <-time.After(5 * time.Second):
```
Comment (r0mant) [inline lib/multiplexer/web.go:108]: Same comment.

File: lib/service/desktop.go [hunk 10]
```diff
@@ -260,14 +260,14 @@ func (process *TeleportProcess) initWindowsDesktopServiceRegistered(log *logrus.
 
 	// Cleanup, when process is exiting.
 	process.OnExit("windows_desktop.shutdown", func(payload interface{}) {
+		warnOnErr(listener.Close(), log)
```
Comment (r0mant) [inline lib/service/desktop.go:263]: If it's important that this goes first, add a comment explaining why for future readers/people who may be refactoring this.

File: lib/service/kubernetes.go [hunk 11]
```diff
@@ -284,9 +284,7 @@ func (process *TeleportProcess) initKubernetesService(log *logrus.Entry, conn *C
 
 	// Cleanup, when process is exiting.
 	process.OnExit("kube.shutdown", func(payload interface{}) {
-		if asyncEmitter != nil {
-			warnOnErr(asyncEmitter.Close(), log)
-		}
+		warnOnErr(listener.Close(), log)
```
Comment (r0mant) [inline lib/service/kubernetes.go:287]: Same as above, I'd add a comment with explanation here and everywhere below where applicable.

File: lib/service/kubernetes.go [hunk 12]
```diff
@@ -298,7 +296,8 @@ func (process *TeleportProcess) initKubernetesService(log *logrus.Entry, conn *C
 			warnOnErr(kubeServer.Close(), log)
 			agentPool.Stop()
 		}
-		warnOnErr(listener.Close(), log)
+
+		warnOnErr(asyncEmitter.Close(), log)
```
Comment (r0mant) [inline lib/service/kubernetes.go:300]: Can `asyncEmitter` no longer be `nil`? The `if asyncEmitter != nil` condition is no longer here.

File: lib/service/signals.go [hunk 13]
```diff
@@ -78,7 +75,6 @@ func (process *TeleportProcess) WaitForSignals(ctx context.Context) error {
 		case signal := <-sigC:
 			switch signal {
 			case syscall.SIGQUIT:
-				go process.printShutdownStatus(doneContext)
```
Comment (r0mant) [inline lib/service/signals.go:81]: Wasn't it useful to see the shutdown status?

File: lib/service/kubernetes.go (hunk 12 above)
Comment (espadolini) [reply lib/service/kubernetes.go]: Reduced PR scope.

File: lib/service/kubernetes.go (hunk 11 above)
Comment (espadolini) [reply lib/service/kubernetes.go]: Reduced PR scope.

File: lib/service/desktop.go (hunk 10 above)
Comment (espadolini) [reply lib/service/desktop.go]: Reduced the scope of this PR to just the proxy, as it's very much a special case and everything else ultimately works fine right now (other than race conditions related to shutting down before or during init, which the current PR wouldn't address anyway). I think there's a better solution for "listener cleanliness" to be managed at the `TeleportProcess` level anyway (by forcibly closing every exported/exportable listener at shutdown).

File: lib/service/kubernetes.go (hunk 12 above)
Comment (espadolini) [reply lib/service/kubernetes.go]: The same comment applies to the `asyncEmitter` in `proxy.shutdown`, which was never nullable in the first place.

**Comment by espadolini** [comment]
@Tener PTAL at 6a73134?

**Comment by espadolini** [comment]
@r0mant PTAL? It would be nice to get this merged before the next Cloud upgrade. I pushed back on the `time.Ticker` vs `time.After` thing but if you feel strongly about it I can just revert those.

**Review by r0mant — APPROVED** [review]

**Pull request merged**